core/src/main/java/org/htsjdk/core/utils/ParamUtils.java

```diff
@@ -0,0 +1,106 @@
+package org.htsjdk.core.utils;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * General utilities to check and validate conditions for constructors/methods.
+ *
+ * <p>All the methods in this class throw {@link IllegalArgumentException} if conditions are not
+ * met.
+ */
+public final class ParamUtils {
+
+    // cannot be instantiated
+    private ParamUtils() {}
+
+    /**
+     * Checks that the condition is {@code true}.
+     *
+     * @param condition required condition.
+     * @param message   the text message that would be passed to the exception thrown.
+     */
+    public static void validate(final boolean condition, final Supplier<String> message) {
+        if (!condition) {
+            throw new IllegalArgumentException(message.get());
+        }
+    }
+
+    /**
+     * Checks whether an index is within the bounds [0, length).
+     *
+     * @param index  query index.
+     * @param length collection or array size.
+     *
+     * @return same value as the input {@code index}.
+     */
+    public static int validateIndex(final int index, final int length) {
+        validate(index >= 0, () -> "index cannot be negative: " + index);
+        validate(index < length, () -> "index out of bounds: " + index + ">" + (length - 1));
+        return index;
+    }
+
+    /**
+     * Checks that an {@link Object} is not {@code null}.
+     *
+     * @param object  any {@link Object}.
+     * @param message the text message that would be passed to the exception thrown.
+     * @param <T>     object type.
+     *
+     * @return the same object (if non-null).
+     *
+     * @throws IllegalArgumentException if a {@code object == null}.
+     */
+    public static <T> T nonNull(final T object, final Supplier<String> message) {
+        if (object != null) {
+            return object;
+        }
+        throw new IllegalArgumentException(message.get());
+    }
+
+    /**
+     * Checks that an {@link Object} is not {@code null}.
+     *
+     * @param object  any {@link Object}.
+     * @param <T>     object type.
+     *
+     * @return the same object (if non-null).
+     *
+     * @throws IllegalArgumentException if a {@code object == null}.
+     */
+    public static <T> T nonNull(final T object) {
+        return nonNull(object, () -> object.getClass() + " cannot be null");

```

Comment:
Won't dereferencing `object` here cause a null pointer?

Reply:
Reason 1 for not coding at 23:00